feat: import formulas when importing xlsx file

AyushAgrawal-A2 opened the pull request:
part of #1435 
ref: https://github.com/quadratichq/quadratic/issues/1435#issuecomment-2179020914

Importing formulas works and results are correct for supported functions.

davidkircos commented:
We should add all Excel formulas to the Rust Formulas parser so that ones that people attempt to use, but are not implemented yet can trigger a mixpanel event and we can prioritize implementation.

AyushAgrawal-A2 commented:
> We should add all Excel formulas to the Rust Formulas parser so that ones that people attempt to use, but are not implemented yet can trigger a mixpanel event and we can prioritize implementation.

@davidkircos 
Changes:
   - Added excel's function list and check for unimplemented functions
   - Added mixpanel tracking for unimplemented error

![image](https://github.com/quadratichq/quadratic/assets/54364088/5a78732f-85e3-4134-a59d-83cc417e6b5a)

Currently, tracking will trigger only when user runs a formula with a valid excel function that is not implemented yet. Importing xlsx file is a server transaction and does not trigger mixpanel tracking. Is this ok ?

ddimaria commented on quadratic-core/src/controller/execution/execute_operation/execute_code.rs:
```diff
@@ -97,8 +97,8 @@ impl GridController {
-            if !transaction.is_user() {
-                unreachable!("Only a user transaction should have a ComputeCode");
+            if !transaction.is_user() && !transaction.is_server() {
+                unreachable!("Only a user/server transaction should have a ComputeCode");
```
I never liked all of this `unreachable` code.  Ideally, all of these failable functions would return results, but that's a lot of restructuring to do.  Could you try using `dbgjs!` instead and avoid the computation in the error case?

AyushAgrawal-A2 replied:
@ddimaria 
Done

ddimaria commented on quadratic-client/src/app/gridGL/cells/CellsArray.ts:
```diff
@@ -64,14 +65,37 @@ export class CellsArray extends Container {
+            });
+          }
+        } catch (error) {
+          console.warn('[CellsArray] Error parsing codeCell.evaluation_result', error);
```
If this is an error case, convert to `console.error` so that Sentry can pick up this error.

AyushAgrawal-A2 replied:
@ddimaria 
Done

ddimaria commented on quadratic-core/src/controller/operations/import.rs:
```diff
@@ -191,12 +199,38 @@ impl GridController {
+                .worksheet_formula(&sheet_name)
+                .map_err(|e: XlsxError| error(e.to_string()))?;
+            let insert_at = formula.start().map_or_else(Pos::default, xlsx_range_to_pos);
+            let mut formula_compute_ops = vec![] as Vec<Operation>;
```
You should be able to drop the `as Vec<Operation>` as this should be inferenced by the push on line 222.

AyushAgrawal-A2 replied:
@ddimaria 
Done

ddimaria commented:
Good PR @AyushAgrawal-A2 !  Can you either add all supported functions to the `imports_a_simple_excel_file()` test located at `src/controller/user_actions/import.rs` or create a new test there for functions?  I think I just generated the initial Excel file from the free online version.

AyushAgrawal-A2 commented:
> Good PR @AyushAgrawal-A2 ! Can you either add all supported functions to the `imports_a_simple_excel_file()` test located at `src/controller/user_actions/import.rs` or create a new test there for functions? I think I just generated the initial Excel file from the free online version.

@ddimaria 
Done

HactarCE commented on quadratic-core/src/controller/operations/import.rs:
```diff
@@ -139,19 +139,31 @@ impl GridController {
+        };
+
+        // remove the _xlfn. _xludf. prefix from the function name
+        let re = Regex::new(r"(?:^|[^A-Za-z\d])(?:_xlfn|_xludf)\.").unwrap();
```
I think this consumes the last character before the `_xlfn.`/`_xludf.`. For example, running the replacement on `1+_xlfn.SUM(A1:10)` results in `1_xlfn.SUM(A1:10)`. Would it be ok to just use `r"(_xlfn|_xludf)\."`?

This also eats any string literal containing these characters, which is admittedly a pathological input but nonetheless may be a reason to properly parse & reserialize formulas when doing the replacement.

AyushAgrawal-A2 replied:
@HactarCE 
I should have used positive look behind `r"(?<=^|[^A-Za-z\d])(?:_xlfn|_xludf)"` to assert a non word character before this prefix. This will not consume other characters.

In first iteration I did not clean the input, instead after lexar tokenization... I used just remove the prefix `_xlfn. / _xludf.` inside `lookup_function`. In this case prefix will always be at the start of string. This approach did not have the issue of removing characters from strings or other places.

I thought it is bad to save a dirty input so switched to cleaning it at import.

AyushAgrawal-A2 replied:
look-ahead is not supported, switched back to earlier implementation. now storing code string as it is.

HactarCE commented on quadratic-core/src/controller/operations/import.rs:
```diff
@@ -130,7 +131,6 @@ impl GridController {
     ) -> Result<Vec<Operation>> {
         let mut ops = vec![] as Vec<Operation>;
-        let insert_at = Pos::default();
         let error =
```
This error is always used with an `XlsxError`, so we can factor the `.to_string()` into here and then simplify the call sites:

```rs
let error = |e: XlsxError| anyhow!("Error parsing Excel file {file_name}: {e}");

// ... then at call sites:
.map_err(error)
```
String formatting implicitly calls `.to_string()`.

HactarCE commented on quadratic-core/src/controller/operations/import.rs:
```diff
@@ -191,12 +203,39 @@ impl GridController {
+            for (y, row) in formula.rows().enumerate() {
+                for (x, cell) in row.iter().enumerate() {
+                    if !cell.is_empty() {
+                        let code = re.replace_all(cell, "").into_owned();
```
Add a comment here explaining what this `.replace_all()` is for, and maybe move the regex closer to the call site. It took me a minute to figure it out because the regex is defined ~65 lines earlier.

You could even put it in a `lazy_static!` _right_ above this line, so that it's only executed once for the whole program but it's still close to where it's used. If that makes the loop body too big, pull out the whole string replacement into a separate function.

HactarCE commented on quadratic-core/src/controller/operations/import.rs:
```diff
@@ -191,12 +203,39 @@ impl GridController {
+                        });
+                        sheet.set_cell_value(pos, cell_value);
+                        // add code compute operation, to generate code runs
+                        formula_compute_ops.push(Operation::ComputeCode {
+                            sheet_pos: pos.to_sheet_pos(sheet.id),
+                        });
```
@ddimaria Do you know if this will work well with formula dependencies? Will it run some code cells multiple times or use outdated results or do any other silly things?

Actually, is this even necessary if we `rerun_all_code_cells_operations()` up in `import_excel()`?

HactarCE commented on quadratic-core/src/controller/user_actions/import.rs:
```diff
@@ -63,6 +69,8 @@ mod tests {
 
     // const EXCEL_FILE: &str = "../quadratic-rust-shared/data/excel/temperature.xlsx";
     const EXCEL_FILE: &str = "../quadratic-rust-shared/data/excel/basic.xlsx";
+    const EXCEL_FUNCTIONS_FILE: &str =
```
Why not [`include_str!()`](https://doc.rust-lang.org/std/macro.include_str.html) and [`include_bytes!()`](https://doc.rust-lang.org/std/macro.include_bytes.html)?

HactarCE commented on quadratic-core/src/controller/user_actions/import.rs:
If not [`include_str!()`](https://doc.rust-lang.org/std/macro.include_str.html), why not [`std::fs::read()`](https://doc.rust-lang.org/std/fs/fn.read.html)?

AyushAgrawal-A2 replied:
I copy-pasted file reading code from other functions in the same file.
switched all to `std::fs::read`

HactarCE reviewed (approved):
Thanks for the changes!

davidkircos commented:
Looks great to me! Going to merge this branch. I repurposed this Issue to track our overall Excel File import support https://github.com/quadratichq/quadratic/issues/1154
